### PR TITLE
Add cartesianProduct to Collection

### LINKED ISSUE #9121: Add new method cartesianProduct to collection class
This is a (multiple allowed):
* [ ] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.3.0

### What you did
Add an new method **cartesianProduct** to collection class, which will perform a cartesian product to the collection.

e.g.

 ```
      $items = [
            ['A', 'B', 'C', 'D'],
            [1, 2, 3, 4]
      ]
     
      $cartesianProduct = (new Collection($items))->cartesianProduct()->toList();
```
     
Returns results as shown below
```
        [
          ['A', 1],
          ['A', 2],
          ['A', 3],
          ['A', 4],
          ['B', 1],
          ['B', 2],
          ['B', 3],
          ['B', 4],
          ['C', 1],
          ['C', 2],
          ['C', 3],
          ['C', 4],
          ['D', 1],
          ['D', 2],
          ['D', 3],
          ['D', 4],
      ]
```

This is a very handy method for processing array data.

### PR DESCRIPTION
**Pull request by hytromo**
To resolve #9121

### CODE REVIEW

File: src/Collection/CollectionTrait.php [hunk 1]
```diff
@@ -686,6 +686,47 @@ public function _unwrap()
      *
      * @return \Cake\Collection\CollectionInterface
      */
+    public function cartesianProduct(callable $operation = null, callable $filter = null)
+    {
+        if ($this->isEmpty()) {
+            return new Collection([]);
```
Comment (markstory) [inline src/Collection/CollectionTrait.php]: Missing a test case for the empty set.

File: src/Collection/CollectionTrait.php [hunk 2]
```diff
@@ -686,6 +686,47 @@ public function _unwrap()
...
+        }
+
+        $result = [];
+        $counts = $this->map(function ($arr) {
+            return count($arr);
+        })->toList();
+        $allArr = $this->toList();
+        $lastIndex = count($allArr) - 1;
+        // holds the indexes of the arrays that generate the current combination
+        $currentIndexes = array_fill(0, $lastIndex + 1, 0);
```
Comment (markstory) [inline src/Collection/CollectionTrait.php]: Isn't `$lastIndex + 1` the same as `count($allArr)`?

File: src/Collection/CollectionTrait.php [hunk 3]
```diff
@@ -686,6 +686,47 @@ public function _unwrap()
...
+    {
+        if ($this->isEmpty()) {
+            return new Collection([]);
+        }
+
+        $result = [];
+        $counts = $this->map(function ($arr) {
+            return count($arr);
+        })->toList();
+        $allArr = $this->toList();
```
Comment (markstory) [inline src/Collection/CollectionTrait.php]: `$values` instead of `$allArr`? `allArr` sounds like pirate speak.

File: src/Collection/CollectionTrait.php [hunk 4]
```diff
@@ -686,6 +686,47 @@ public function _unwrap()
...
+        $currentIndexes = array_fill(0, $lastIndex + 1, 0);
+        $changeIndex = $lastIndex;
+
+        while (!($changeIndex === 0 && $currentIndexes[0] === $counts[0])) {
+            $currentCombination = array_map(function ($arr, $index) {
+                return $arr[$index];
+            }, $allArr, $currentIndexes);
+
+            if ($filter === null || $filter(...$currentCombination)) {
+                $result[] = ($operation === null) ? $currentCombination : $operation(...$currentCombination);
```
Comment (markstory) [inline src/Collection/CollectionTrait.php]: You can't use spread operators in 3.3.x, this will have to target 3.4.x if you want to use the spread operator.

File: tests/TestCase/Collection/CollectionTest.php [hunk 5]
```diff
@@ -1770,6 +1770,100 @@ public function testChunkNested()
...
+            [3 => ['C' => 'telephone']],
+            [4 => ['A' => 'name']],
+            [4 => ['A' => 'telephone']],
+            [4 => ['B' => 'name']],
+            [4 => ['B' => 'telephone']],
+            [4 => ['C' => 'name']],
+            [4 => ['C' => 'telephone']]
+        ];
+
+        $this->assertEquals($expected, $result->toList());
```
Comment (markstory) [inline tests/TestCase/Collection/CollectionTest.php]: What about multi-dimensional contents? What happens with those?

File: src/Collection/CollectionTrait.php (hunk 2 above)
Comment (hytromo) [inline src/Collection/CollectionTrait.php]: Yes. I preferred to call `count()` once. `$lastIndex` is used inside the loop.

So, have to choose between

- Call `count()` mutliple times

- Use `$lastIndex + 1` (as it is now)

- Use `$count = count(..)` and then use `$count - 1` where needed

**Comment by hytromo** [comment]
So, I agree that this should work somehow for multidimensional arrays, but I cannot figure out what would one expect.

**Comment by hytromo** [comment]
I propose the following.

Let collection be

```
$collection = new Collection([
    [
        'Admins' => 
        [
            [
                'name' => 'admin',
                'email' => 'admin@example.com'
            ],
            [
                'name' => 'root',
                'email' => 'root@example.com'
            ]
        ]
    ],
    [
        'Users' => 
        [
            [
                'name' => 'alex',
                'email' => 'alex@example.com'
            ],
            [
                'name' => 'mark',
                'email' => 'mark@example.com'
            ]
        ]
    ]
]);
```

then something like this:

---
`$collection->cartesianProduct('{*}.name');`

will produce the cartesian product of all the inner name values, e.g.

```
[
    ['admin', 'alex'],
    ['admin', 'mark'],
    ['root', 'alex'],
    ['root', 'mark'],
]
```
---
`$result = $collection->cartesianProduct(['{*}.name', '{*}.email']);`

will produce the cartesian product of all the inner name values with all the inner email values, e.g.

```
[
    ['admin', 'admin@example.com'],
    ['admin', 'root@example.com'],
    ['admin', 'alex@example.com'],
    ['admin', 'mark@example.com'],
    ['root', 'admin@example.com'],
    ['root', 'root@example.com'],
    ['root', 'alex@example.com'],
    ['root', 'mark@example.com'],
    ['alex', 'admin@example.com'],
    ['alex', 'root@example.com'],
    ['alex', 'alex@example.com'] 
    ['alex', 'mark@example.com'],
    ['mark', 'admin@example.com'],
    ['mark', 'root@example.com'],
    ['mark', 'alex@example.com'],
    ['mark', 'mark@example.com'],
]
```

Notice that this matches names and emails from the same array (in the previous example the names were matched from the different arrays). I think this should be the expected behavior

---

Then, I think it would be nice to make the user able to extract different things from different arrays, so

```
$result = $collection->cartesianProduct([[0 => '{*}.name'], '{*}.email']);
// or
$result = $collection->cartesianProduct([['{*}.name'], '{*}.email']);
```

will produce the cartesian product of all the inner name values of the first array with all the inner email values (of both arrays), e.g.

```
    ['admin', 'admin@example.com'],
    ['admin', 'root@example.com'],
    ['admin', 'alex@example.com'],
    ['admin', 'mark@example.com'],
    ['root', 'admin@example.com'],
    ['root', 'root@example.com'],
    ['root', 'alex@example.com'],
    ['root', 'mark@example.com'],
]
```
---

Another example of this would be

```
$result = $collection->cartesianProduct([[0 => '{*}.name', 1 => '{*}.email']]);
// or
$result = $collection->cartesianProduct([['{*}.name', '{*}.email']]);
```

which generates the cartesian product of the names of the first array with the emails of the second one, e.g.

```
[
    ['admin', 'alex@example.com'],
    ['admin', 'mark@example.com'],
    ['root', 'alex@example.com'],
    ['root', 'mark@example.com'],
]
```
---

Finally, the user will be able to combine the matcher with the `$operation` and `$filter` closures, e.g.

```
$result = $collection->cartesianProduct([[0 => '{*}.name', 1 => '{*}.email']], function ($name, $email) {
    return ['name' => $name, 'email' => $email];
});
```

the above will produce the same thing with the previous example but with the keys `name` and `email`:

```
[
    [
        'name' => 'admin',
        'email' => 'alex@example.com'
    ],
    [
        'name' => 'admin',
        'email' => 'mark@example.com'
    ],
    [
        'name' => 'root',
        'email' => 'alex@example.com'
    ],
    [
        'name' => 'root',
        'email' => 'mark@example.com'
    ],
]
```

**Comment by markstory** [comment]
Another option is to fail on multi-dimensional arrays as it is not realistic to generate sane cartesian products from them.

**Comment by hytromo** [comment]
Yeah simply calling for cartesian product in a multiD array doesn't make sense.

But don't you think that it will be useful to call it with a matcher (as shown in the examples above) so as to allow the user use it directly with collections containing result sets from the db?

As long as the user has specified a matcher then the product does make sense.

**Comment by markstory** [comment]
I think the matcher could make operating on multi-dimensional arrays possible, but I worry about taking on the additional complexity when I'm not sure how frequent doing Cartesian products of multi-dimensional arrays will be.

File: src/Collection/CollectionTrait.php [hunk 6]
```diff
@@ -692,23 +692,38 @@ public function cartesianProduct(callable $operation = null, callable $filter =
...
 
+        // contains each array of the collection
+        $values = [];
+        // contains the keys of each array
+        $valuesKeys = [];
+        // contains the number of elements of each array
+        $counts = [];
+
+        foreach ($this as $value) {
+            // fail if any of the values is a multidimensional array
```
Comment (dereuromark) [inline src/Collection/CollectionTrait.php]: I dont think all the inline comments are necessary, with a good exception etc here
also the ones above should be clear from the variable names ideally.

File: tests/TestCase/Collection/CollectionTest.php [hunk 7]
```diff
@@ -1858,10 +1866,57 @@ public function testCartesianProduct()
...
+            ['kostas', 'kostas@example.com'],
+            ['kostas', 'leon@example.com'],
+            ['leon', 'alex@example.com'],
+            ['leon', 'kostas@example.com'],
+            ['leon', 'leon@example.com'],
         ];
 
         $this->assertEquals($expected, $result->toList());
+
+        $this->expectException(\LogicException::class);
```
Comment (markstory) [inline tests/TestCase/Collection/CollectionTest.php]: This error case could be its own test method.

File: src/Collection/CollectionTrait.php [hunk 8]
```diff
@@ -686,6 +686,58 @@ public function _unwrap()
...
+        if ($this->isEmpty()) {
+            return new Collection([]);
+        }
+
+        $collectionArrays = [];
+        $collectionArraysKeys = [];
+        $collectionArraysCounts = [];
+
+        foreach ($this as $value) {
+            if (count($value) !== count($value, COUNT_RECURSIVE)) {
```
Comment (lorenzo) [inline src/Collection/CollectionTrait.php]: How can you be sure that `$value` is countable?

File: src/Collection/CollectionTrait.php [hunk 9]
```diff
@@ -686,6 +686,58 @@ public function _unwrap()
...
+        $currentIndexes = array_fill(0, $lastIndex + 1, 0);
+
+        $changeIndex = $lastIndex;
+
+        while (!($changeIndex === 0 && $currentIndexes[0] === $collectionArraysCounts[0])) {
+            $currentCombination = array_map(function ($value, $keys, $index) {
+                return $value[$keys[$index]];
+            }, $collectionArrays, $collectionArraysKeys, $currentIndexes);
+
+            if ($filter === null || call_user_func_array($filter, $currentCombination)) {
```
Comment (lorenzo) [inline src/Collection/CollectionTrait.php]: I think making each value in the combination a function parameter makes this really difficult to use. Why not just pass an array as a single argument to the function?

**Comment by lorenzo** [comment]
Thanks for the great work @hytromo
